packages/rest/src/BaseResource.ts

```diff
@@ -0,0 +1,263 @@
+    name: string,
+    construct: () => T,
+  ): T {
+    if (!Object.hasOwnProperty.call(this, this.cacheSymbol)) {
+      (this as any)[this.cacheSymbol] = {};
+    }
+    const cache = (this as any)[this.cacheSymbol];
+    if (!(name in cache)) {
+      // eslint-disable-next-line @typescript-eslint/no-this-alias
+      const resource: any = this;
+      cache[name] = construct().extend({
+        get name() {
+          return `${resource.name}.${name.replace(/#/, '')}`;
+        },
+      });
+    }
+    return cache[name].useFetchInit() as T;
+  }
+
+  /** Base endpoint that uses all the hooks provided by Resource  */
+  protected static endpoint(): RestEndpoint<
+    (this: RestEndpoint, params: any) => Promise<any>,
+    Schema | undefined,
+    undefined
+  > {
+    return this.memo('#endpoint', () => {
+      // eslint-disable-next-line
+      const resource = this;
+      const instanceFetch = this.fetch.bind(this);
+      const url = this.url.bind(this);
+
+      return new Endpoint(
+        function (params: any) {
+          return instanceFetch(this.url(params), this.getFetchInit());
+        },
+        {
+          ...this.getEndpointExtra(),
+          key: function (this: any, params: any) {
+            return `${this.method} ${this.url(params)}`;
+          },
+          url,
+          fetchInit: {} as RequestInit,
+          useFetchInit(this: any) {
+            this.fetchInit = resource.useFetchInit(this.fetchInit);
+            return this;
+          },
+          getFetchInit(this: any, body?: any) {
+            if (isPojo(body)) {
+              body = JSON.stringify(body);
+            }
+            return resource.getFetchInit({
+              ...this.fetchInit,
+              method: this.method,
+              signal: this.signal,
+              body,
+            });
+          },
+          method: 'GET',
+          signal: undefined as AbortSignal | undefined,
+        },
+      );
+    });
+  }
+
+  /** Base endpoint but for sideEffects */
+  protected static endpointMutate(): RestEndpoint<
+    (this: RestEndpoint, params: any, body?: any) => Promise<any>,
+    Schema | undefined,
+    true
+  > {
+    const instanceFetch = this.fetch.bind(this);
+    const endpoint = this.endpoint();
+    return this.memo('#endpointMutate', () =>
+      endpoint.extend({
+        fetch(this: RestEndpoint, params: any, body: any) {
+          return instanceFetch(this.url(params), this.getFetchInit(body));
+        },
+        sideEffect: true,
+        method: 'POST',
+      }),
+    );
+  }
+
+  static {
+    Object.defineProperty(this.prototype, 'url', {
+      get(): string {
+        // typescript thinks constructor is just a function
+        const Static = this.constructor as typeof BaseResource;
+        return Static.url(this);
+      },
+      set(v: string) {
+        Object.defineProperty(this, 'url', {
+          value: v,
+          writable: true,
+          enumerable: true,
+        });
+      },
+    });
```

Comment:
why can't these be normal prototype getter methods again?

Reply:
TypeScript differentiates between getters and non getters so it doesn't let people define normal url properties which is what this setter is supposed to allow

Reply:
and there's no way to use TS comments to fix it, instead of writing code that results in a larger bundle?

Reply:
what is a TS comment?

Reply:
https://github.com/coinbase/rest-hooks/pull/1737#discussion_r815360548 < original conversation

Reply:
Another key distinction is the setter/getter are on the prototype, but the setter sets a property on the actual value itself, which overrides this by avoiding the prototype chain lookup.

Reply:
oh sure, but `set url(v) {  Object.defineProperty(…); }` would achieve the same thing, if not for the TS stuff.